geodataRetriever.py

```diff
@@ -58,72 +77,110 @@ def SendEmail(subject, message, TO = toEmails, FROM = fromEmail, password = from
     except Exception:
         logging.exception("Could not send email")
 
+# Update data in ArcGIS Workspace
+def UpdateFeatureClass(file, fcName):
+    print("Updating feature class or table...")
+
+    # if CSV, get fields
+    if file.endswith('.csv'):
+        cols = []
+        with open(file, encoding="utf-8-sig") as f:
+            reader = csv.reader(f)
+            cols = next(reader)
+
+        # Remove FID, OBJECTID
+        fields = []
+        for f in cols:
+            if f != 'FID' and f != 'OBJECTID':
+                f = re.sub(r'\W+', '_', f)  # replace non-alphanumeric characters with underscore
+                fields.append(f) 
+
+        # Check that all fields in csv are in target table
+        fcFields = [f.name for f in arcpy.ListFields(fcName) if f.type != 'OID' and f.type != 'Geometry']
+        if not set(fields).issubset(set(fcFields)):
+            print("ERROR: Fields in CSV do not match fields in target table")
+            logging.error("Fields in CSV do not match fields in target table " + fcName)
+            SendEmail('Geodata Retriever failed', "Fields in downloaded CSV do not match fields in target table" + fcName + "\nSee log at " + logFilePath)
+            return
+
+    # if SHP, get fields
+    if file.endswith('.shp'):
+        # Get field names 
+        shapefileFields = [f.name for f in arcpy.ListFields(file) if f.type != 'OID' and f.type != 'Geometry']
+        fcFields = [f.name for f in arcpy.ListFields(fcName) if f.type != 'OID' and f.type != 'Geometry']
+
+        # Remove FID, OBJECTID
+        fields = []
+        for f in shapefileFields:
+            if f != 'FID' and f != 'OBJECTID':
+                fields.append(f)
+
+        # Check that all fields in shapefile are in target feature class
+        if not set(fields).issubset(set(fcFields)):
```

Comment:
What happens here if the csv has a blank row before the data starts? Does this fail gracefully or crash?

Reply:
That should be plenty